File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/strategy/solution/packer.cpp
// packer.cpp

#include "dp_exception.h"
#include "packer.h"
#include "packing_strategy.h"

packer::packer(packing_strategy * a_strategy)
{
	set_strategy(a_strategy);
}

void packer::pack()
{
	get_strategy()->pack(*this);
}

void packer::set_strategy(packing_strategy * a_strategy)
{
	if (a_strategy == nullptr)
	{
		throw dp_exception("void packer::set_strategy(packing_strategy * a_strategy) -- Missing packing_strategy.");
	}

	my_strategy = a_strategy;
}

packing_strategy * packer::get_strategy() const
{
	return my_strategy;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/strategy/solution/Packer.java
// Packer.java

public class Packer
{
	public Packer(PackingStrategy aStrategy)
	{
		setStrategy(aStrategy);
	}

	public void pack()
	{
		getStrategy().pack(this);
	}

	public void setStrategy(PackingStrategy aStrategy)
	{
		if (aStrategy == null)
		{
			throw new DpException("Missing strategy.");
		}
		myStrategy = aStrategy;
	}

	public PackingStrategy getStrategy()
	{
		return myStrategy;
	}

	private PackingStrategy myStrategy;
}
